File: core/src/com/desitum/library/view/SeekBar.java
package com.desitum.library.view;

import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.desitum.library.drawing.Drawable;
import com.desitum.library.game.World;

/**
 * Created by kodyvanry on 5/1/17.
 */

public class SeekBar extends ProgressBar {

    private static final float SEEKER_SIZE = DEFAULT_PROGRESS_BAR_HEIGHT * 2;

    private Drawable mSeekerDrawable;
    private OnSeekBarChangeListener mOnSeekBarChangeListener;

    public SeekBar(World world) {
        super(world);
    }

    public SeekBar(World world, LayoutConstraints layoutConstraints) {
        super(world, layoutConstraints);
    }

    @Override
    public boolean onTouchEvent(TouchEvent touchEvent) {
        switch (touchEvent.getAction()) {
            case DOWN:
                setProgress((touchEvent.getX() - getX()) / getWidth());
                if (mOnSeekBarChangeListener != null) {
                    mOnSeekBarChangeListener.onStartTrackingTouch(this);
                    mOnSeekBarChangeListener.onSeekChanged(this, getProgress());
                }
                break;
            case MOVE:
                setProgress((touchEvent.getX() - getX()) / getWidth());
                if (mOnSeekBarChangeListener != null)
                    mOnSeekBarChangeListener.onSeekChanged(this, getProgress());
                break;
            case UP:
                if (mOnSeekBarChangeListener != null)
                    mOnSeekBarChangeListener.onStopTrackingTouch(this);
        }
        return super.onTouchEvent(touchEvent);
    }

    @Override
    public void onDraw(Batch batch, Viewport viewport) {
        super.onDraw(batch, viewport);
        if (mSeekerDrawable != null)
            mSeekerDrawable.draw(
                    batch,
                    getX() + getWidth() * getProgress() - SEEKER_SIZE / 2,
                    getY() + getHeight() / 2 - SEEKER_SIZE / 2,
                    SEEKER_SIZE,
                    SEEKER_SIZE
                    );
    }

    public void setSeekerDrawable(Drawable seekerDrawable) {
        this.mSeekerDrawable = seekerDrawable;
    }

    public void setOnSeekBarChangeListener(OnSeekBarChangeListener onSeekBarChangeListener) {
        this.mOnSeekBarChangeListener = onSeekBarChangeListener;
    }

    public interface OnSeekBarChangeListener {
        void onSeekChanged(View view, float value);

        void onStartTrackingTouch(View view);

        void onStopTrackingTouch(View view);
    }
}

File: core/src/com/desitum/library/view/SeekBar.kt
package com.desitum.library.view

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.utils.viewport.Viewport
import com.desitum.library.drawing.Drawable
import com.desitum.library.game.World

/**
 * Created by kodyvanry on 5/1/17.
 */

open class SeekBar : ProgressBar {

    var seekerDrawable: Drawable? = null
    private var onSeekBarChangeListener: OnSeekBarChangeListener? = null
    private var seekerSize = SEEKER_SIZE

    constructor(world: World) : this(world, null)

    constructor(world: World, layoutConstraints: LayoutConstraints?) : super(world, layoutConstraints)

    override fun onTouchEvent(touchEvent: TouchEvent): Boolean {
        when (touchEvent.action) {
            TouchEvent.Action.DOWN -> {
                progress = (touchEvent.x - x) / width
                onSeekBarChangeListener?.let {
                    it.onStartTrackingTouch(this)
                    it.onSeekChanged(this, progress)
                }
            }
            TouchEvent.Action.MOVE -> {
                progress = (touchEvent.x - x) / width
                onSeekBarChangeListener?.onSeekChanged(this, progress)
            }
            TouchEvent.Action.UP -> onSeekBarChangeListener?.onStopTrackingTouch(this)
            else -> {}
        }
        return super.onTouchEvent(touchEvent)
    }

    override fun onDraw(batch: Batch, viewport: Viewport) {
        super.onDraw(batch, viewport)
        seekerDrawable?.draw(
                    batch,
                    x + width * progress - seekerSize / 2,
                    y + height / 2 - seekerSize / 2,
                seekerSize,
                seekerSize
            )
    }

    interface OnSeekBarChangeListener {
        fun onSeekChanged(view: View, value: Float)

        fun onStartTrackingTouch(view: View)

        fun onStopTrackingTouch(view: View)
    }

    companion object {

        private val SEEKER_SIZE = ProgressBar.DEFAULT_PROGRESS_BAR_HEIGHT * 2
    }
}
